### PR TITLE
[EditorPreview] Prevent previewing tokens out of scope of the selected frame

### PR DESCRIPTION
**Pull request by ryanjduffy**
Associated Issue: #2398

### Summary of Changes

* Add `resolveToken()` to parser worker which resolves a token to its expression and determines if the token is in scope of the given frame
* Replace `getExpressionFromToken()` in `utils/Editor` with `resolveToken`

### Test Plan

- [x] Set a breakpoint within a function
- [x] Trigger the breakpoint
- [x] Hovering tokens within the function should preview
- [x] Hovering tokens outside the function or in another file should not preview
- [x] Clicking "x" closes the panel

### Screenshots/Videos (OPTIONAL)
debugger html-2398

### CODE REVIEW

File: src/utils/parser/utils.js [hunk 1]
```diff
@@ -228,24 +255,60 @@ function getExpression(
...
+    enter(path) {
+      // if we haven't found an expression yet, determine if the token is part
+      // of one
+      if (!expression) {
+        expression = resolveExpression(path, token, location);
+      }
+
+      // determine if the current path is a function or program containing the
+      // frame
+      const scope = resolveScope(path, frame.location);
```
Comment (jasonLaster) [inline src/utils/parser/utils.js]: could we do 

```js
let scope = null
```

and then below:

```js
scope = resolveScope(path, frame.location);
```

File: src/utils/parser/utils.js [hunk 2]
```diff
@@ -228,24 +255,60 @@ function getExpression(
...
+
+  if (isEmpty(ast)) {
+    return;
+  }
+
+  traverse(ast, {
+    enter(path) {
+      // if we haven't found an expression yet, determine if the token is part
+      // of one
+      if (!expression) {
```
Comment (jasonLaster) [inline src/utils/parser/utils.js]: why `if (!expression) {`

File: src/utils/parser/utils.js [hunk 3]
```diff
@@ -228,24 +255,60 @@ function getExpression(
...
+        scopes.unshift(scope);
+      }
+    },
+  });
+
+  let inScope = false;
+  if (scopes.length > 0) {
+    // determine if the narrowest scope contains the token's location
+    inScope = nodeContainsLocation({ node: scopes[0].node, location });
+  }
```
Comment (jasonLaster) [inline src/utils/parser/utils.js]: hmm, could we get to this point? we should always have a scope, no?

```js
const inScope = nodeContainsLocation({ node: scope.node, location })
```

File: src/components/Editor/index.js [hunk 4]
```diff
@@ -313,30 +319,33 @@ const Editor = React.createClass({
       this.setState({ selectedToken: null, selectedExpression: null });
     }
 
-    const expressionFromToken = await getExpressionFromToken(
+    const resolution = await resolveToken(
```
Comment (jasonLaster) [inline src/components/Editor/index.js]: destructuring ftw 

```js
const { expression, inScope } = await resolveToken(
```

File: src/components/Editor/index.js [hunk 5]
```diff
@@ -313,30 +319,33 @@ const Editor = React.createClass({
...
+      sourceText,
+      selectedFrame
     );
 
-    if (expressionFromToken) {
-      addExpression(expressionFromToken.value, { visible: false });
-    }
-
-    const variables = getVisibleVariablesFromScope(pauseData, selectedFrame);
+    if (resolution.inScope) {
```
Comment (jasonLaster) [inline src/components/Editor/index.js]: could we bail early?

```js
if (!inScope) {
return;
}
```

File: src/utils/parser/utils.js (hunk 2 above)
Comment (ryanjduffy) [inline src/utils/parser/utils.js]: There's no need to continue looking for an `expression` if we've found one, right? Maybe I'm misunderstanding how the AST traversal happens, though.

File: src/utils/parser/utils.js (hunk 1 above)
Comment (ryanjduffy) [inline src/utils/parser/utils.js]: Sure. Do you prefer all the variables declared at the top of the function?

File: src/utils/parser/utils.js (hunk 3 above)
Comment (ryanjduffy) [inline src/utils/parser/utils.js]: We should. I'd assume there would always be a `Program` scope but I was being defensive. I can remove the check.

File: src/components/Editor/index.js (hunk 5 above)
Comment (ryanjduffy) [inline src/components/Editor/index.js]: Yup

**Comment by ryanjduffy** [comment]
@jasonLaster Updated with code style fixes and tests for `resolveToken`. Do you think I should remove `getExpression()` since it is no longer used (as far as I can tell, at least)? Do you have future plans for it?

**Comment by jasonLaster** [comment]
💥  Awesome!

**Pull request merged**